src/_pytest/cacheprovider.py

```diff
@@ -50,7 +50,9 @@ def for_config(cls, config):
 
     @staticmethod
     def cache_dir_from_config(config):
-        return resolve_from_str(config.getini("cache_dir"), config.rootdir)
+        return resolve_from_str(
+            config.getini("cache_dir"), str(config.rootdir)
+        )
```

Comment:
Instead of casting to string and back to `Path`, `resolve_from_str` should be changed:
```diff
diff --git i/src/_pytest/cacheprovider.py w/src/_pytest/cacheprovider.py
index d5746c3ea..38e02c55a 100755
--- i/src/_pytest/cacheprovider.py
+++ w/src/_pytest/cacheprovider.py
@@ -50,9 +50,7 @@ def for_config(cls, config):
 
     @staticmethod
     def cache_dir_from_config(config):
-        return resolve_from_str(
-            config.getini("cache_dir"), str(config.rootdir)
-        )
+        return resolve_from_str(config.getini("cache_dir"), config.rootdir)
 
     def warn(self, fmt, **args):
         from _pytest.warnings import _issue_warning_captured
diff --git i/src/_pytest/pathlib.py w/src/_pytest/pathlib.py
index 3f65bace1..3d7866319 100644
--- i/src/_pytest/pathlib.py
+++ w/src/_pytest/pathlib.py
@@ -252,9 +252,7 @@ def make_numbered_dir_with_cleanup(root, prefix, keep, lock_timeout):
     raise e
 
 
-def resolve_from_str(input, root):
-    assert not isinstance(input, Path), "would break on py2"
-    root = Path(root)
+def resolve_from_str(input, root: Path) -> Path:
     input = expanduser(input)
     input = expandvars(input)
     if isabs(input):
```

Reply:
But it changes the API after all.. feel free to ignore it then.

Reply:
Indeed, but I will fix the comment, thanks for pointing it out. 👍

Reply:
since we break api on old python if we do this, i believe we should just plan for dropping the dependency together with python3.5

Reply:
Which dependency you mean @RonnyPfannschmidt? Note that this PR already drops `pathlib2`.